BLD: Add Accelerate support for macOS 13.3+

thalassemia opened the pull request:
I added a script ([scipy/_build_utils/_generate_blas_wrapper.py](https://github.com/scipy/scipy/compare/main...thalassemia:accelerate?expand=1#diff-b9fb73b1113defdbecdd9276cd1a893eac534c6b47a53e3512195ee85a2f6203)) to generate C wrappers for Accelerate BLAS/LAPACK. These wrappers are compiled into a static library that other modules can link against with no code modifications (e.g. a call to `dgemm` in a Fortran module will be routed through a wrapper and turned into a call to `dgemm$NEWLAPACK`). I also added Accelerate support to the script ([scipy/linalg/_generate_pyx.py](https://github.com/scipy/scipy/compare/main...thalassemia:accelerate?expand=1#diff-23487f1e98e4e3ec6489ff8fa3919ba87494ddfb3dea4af817b9c61de6c41c52)) that generates the Cython for `_fblas`/`_flapack` using similar C wrappers. In the process of making these two changes, I included special logic to handle the G77 ABI wrappers, eliminating the need for additional Fortran wrappers.

Other changes:
- Functions in `_fblas`/`_flapack` that have a return value were written as such (previously, f2py generated Fortran wrappers for these)
  
  - UPDATE: Reverting this change because the intermediate Fortran wrappers are needed to guarantee compatibility of complex return types, without which segfaults can occur on x86 machines (#19855)
  - UPDATE 2: Brought this change back to avoid [erroneous F2PY boilerplate](https://github.com/scipy/scipy/pull/19816#issuecomment-1970338432) that referenced incorrect BLAS/LAPACK symbols. Fortran subroutine wrappers were translated into C and appended to G77 ABI wrapper file (#20232)

- Resolved missing prototype build warnings in [scipy/integrate/lsoda.pyf](https://github.com/scipy/scipy/compare/main...thalassemia:accelerate?expand=1#diff-bcfb240e4cf0b9546fdc187135f5e1711944e85a992c31e5b8057d62b3d10819) and [scipy/integrate/vode.pyf](https://github.com/scipy/scipy/compare/main...thalassemia:accelerate?expand=1#diff-3a625b9354fc6062bb9c6c0d28d527e376da21034e7cd296e8754afb63cf9051)

  - UPDATE: Separated into #19821
 
- Added test for `cython_blas.cladiv` and `cython_blas.dladiv`

  - UPDATE: Separated into #19820
 
- Declared separate `blas_dep` and `lapack_dep` with necessary compile arguments, leaving `blas` and `lapack` untouched for more useful information when calling `scipy.show_config()`
- Resolved function with no return value build warnings and remove custom G77 wrappers in PROPACK (depends on [separate PR](https://github.com/scipy/PROPACK/pull/7) on the PROPACK repo)

  - UPDATE: Separated into #19855

#### Downsides
The wrapper functions are all in separate source files, allowing the linker to exclude unused symbols and keep binary sizes about the same as they currently are. Since there are ~1600 BLAS/LAPACK functions, this increases build time by a noticeable amount. I tried writing everything to a single file and compiling with `-ffunction-sections -fdata-sections -Wl,-dead_strip`, but this still resulted in unused symbols and larger binaries.

There is no way to mix and match BLAS/LAPACK libraries at the moment (also a challenge to support both LP64 and ILP64).

#### Future
This wrapper generation framework makes it pretty easy to add ILP64 support. If this PR seems OK in principle, I can make a separate PR detailing some quirks I encountered trying to implement ILP64 support with this as a base.

thalassemia committed:
BLD: Add Accelerate support for macOS 13.3+

Generate C wrappers that route F_FUNC(name,NAME)
calls to BLAS_FUNC(name)

ilayn commented:
Thanks @thalassemia this looks like a lot of effort so thank you for that. My initial impression is that this is way too Accelerate specific and treats the design as "everything else vs Accelerate". I think if we are to handle this we have to make Accelerate the special case. 

Could you please elaborate on exactly what Accelerate needs to work and what we are not providing? A mailing list post is also fine if you prefer to write there for a broader audience.

As a side note, I don't know what to think about Accelerate yet; we have bitten by it before massively and I am really hesitant to re-enable it regardless of what Apple paints itself to be this time around. Last time it took a few years to arrive at a conclusion and here is the resulting discussion https://github.com/scipy/scipy/wiki/Dropping-support-for-Accelerate

andyfaff commented:
Also relevant is that @rgommers is working on re-enabling Accelerate support. I believe most of the roadblocks that we were experiencing have been ameliorated on macOS >= Sonoma.X.

Having said that, the rest of Ilhan's points still apply.

thalassemia commented:
We are missing a way to link the BLAS/LAPACK calls scattered throughout the codebase to the non-standard symbols provided by Accelerate. These non-standard symbols are of the format `{lowercase name}$NEWLAPACK{$ILP64}`. This is problematic because many Fortran compilers mangle names by appending an underscore and converting everything to lowercase. This prevents us from simply using direct string substitutions in the relevant Fortran code. 

I considered writing Fortran interfaces with `bind(c)` and `iso_c_binding` but deemed it too much of a hassle since `cython_blas/lapack_signatures.txt` lacks the dimensions of array arguments, something I would need to generate the interfaces. Additionally, SuperLU, a C library, also links against BLAS/LAPACK, so I thought going with C wrappers was the better idea all around.

IMO, this PR already treats Accelerate as an exception to the rule, setting extra flags in `meson.build` files that funnel the wrapper-generating scripts into specific `if` blocks when it is requested. The build simply skips much of the new code when a BLAS/LAPACK library with standard symbols (e.g. LP64 OpenBLAS) is requested. Additionally, this extra logic to handle unconventional symbol names can be easily adapted to add ILP64 support across the board, providing a way to append `64_` to all BLAS/LAPACK calls.

Thanks for the link to the Accelerate discussion. In light of Apple's recent overhaul of Accelerate to align with LAPACK 3.9.1 and include an ILP64 interface, I think that now is a good time to bring back support just as Numpy recently did. The performance advantage for certain operations is fairly compelling.

andyfaff commented:
When we re-add Accelerate support any existing functionality will be available. Ralf will be able to chip in here, but I think his work is going to take care of all the name mangling issues.

ilayn commented:
> In light of Apple's recent overhaul of Accelerate to align with LAPACK 3.9.1 and include an ILP64 interface, I think that now is a good time to bring back support just as Numpy recently did. 

Note that 3.9.1 is from 3 three years ago already, and this precise reason got us into that previous mess. I don't want to commit to the Accelerate framework. Or put differently, we should not be the compatibility seeker and if things break for Accelerate users they should bother Apple and not us, (in English, it won't get fixed). This blocked SciPy for years already. I don't want this to hamper SciPy. It has already happened once and I see no reason for trusting Apple again. So if it works with Accelerate that's awesome but if it doesn't work then we should not be forced to *support* it. 

> The performance advantage for certain operations is fairly compelling.

The performance comes from BLAS and not LAPACK parts. LAPACK is pretty much high level code that uses BLAS ops. Hence if BLAS gets faster so do LAPACK funcs.

In the meanwhile, C++ standard accepted std::linalg for C++26, so, while there is so many things to complain about C++, I'll be the first one to hop on board to that train if it materializes.

https://www.youtube.com/watch?v=-UXHMlAMXNk

ilayn commented:
As a separate comment, PROPACK will be rewritten in Python parts for v.1.13 hence that won't be too much of a problem for BLAS/LAPACK inclusions.

rgommers commented:
Thanks a lot for this PR @thalassemia, this will be really helpful!

Let me first address the below, and then do some testing and come back with more in-depth review comments.

> Note that 3.9.1 is from 3 three years ago already, and this precise reason got us into that previous mess. I don't want to commit to the Accelerate framework. Or put differently, we should not be the compatibility seeker and if things break for Accelerate users they should bother Apple and not us, (in English, it won't get fixed). This blocked SciPy for years already. I don't want this to hamper SciPy. It has already happened once and I see no reason for trusting Apple again. So if it works with Accelerate that's awesome but if it doesn't work then we should not be forced to support it.

I was just working on the exact same thing as @thalassemia, because:
- Accelerate is way faster than OpenBLAS (see, e.g., benchmarks on https://github.com/numpy/numpy/pull/24053),
- Accelerate would allow us to ship way smaller wheels for macOS >=14 (NumPy wheels got ~3x smaller, see https://github.com/numpy/numpy/pull/25255)
- all known problems we had are solved,
- importantly, we now get responses from Apple and even active engineering help (BLAS/LAPACK, and also with compiler issues): https://github.com/numpy/numpy/pulls?q=is%3Apr+developer-ecosystem-engineering+author%3Adeveloper-ecosystem-engineering+is%3Aclosed

In case we find ourselves in a situation like before again several years down the line, we can always drop support again. But right now there are only upsides to supporting Accelerate, and I don't see any reason not to.

Note that this PR looks complex, but there isn't much in it that we don't also need in some form for ILP64 support.

rgommers commented:
After some initial testing it looks like everything works as advertised with Accelerate. The binary size looks fine (only a ~30 kb increase from the wrappers, while dropping vendoring of OpenBLAS gains ~7 MB).

The segfault on the 32-bit Linux CI job looks real.

> I created union types to perform the conversion assuming that Numpy and Cython use the same memory structure for their complex types. I don't know whether it is safe to make this assumption or how else to handle this.

Yes, memory layout is the same and using unions is a common way to deal with complex types.

> Resolved function with no return value build warnings and remove custom G77 wrappers in PROPACK (depends on [separate PR](https://github.com/scipy/PROPACK/pull/7) on the PROPACK repo)

This is really nice. I actually thought that it wasn't possible - but my memory is very hazy there, and if it works it works!



> Functions in `_fblas`/`_flapack` that have a return value were written as such (previously, f2py generated Fortran wrappers for these)
> 
> Resolved missing prototype build warnings in [scipy/integrate/lsoda.pyf](https://github.com/scipy/scipy/compare/main...thalassemia:accelerate?expand=1#diff-bcfb240e4cf0b9546fdc187135f5e1711944e85a992c31e5b8057d62b3d10819) and [scipy/integrate/vode.pyf](https://github.com/scipy/scipy/compare/main...thalassemia:accelerate?expand=1#diff-3a625b9354fc6062bb9c6c0d28d527e376da21034e7cd296e8754afb63cf9051)
> 
> Added test for `cython_blas.cladiv` and `cython_blas.dladiv`

These are all useful changes, thanks. Would you be able to submit these as separate PRs? Then we can get those merged straight away, which is helpful both in itself and to reduce the size of the diff in this PR. 

The removal of the `wrap_g77_abi_*` wrappers could also be done independently of the code for BLAS symbol redirection, right? If so, that'd be great to do as a separate PR too. It can then still be tested merged with this one, and also with MKL (uses F77 ABI too).

> There is no way to mix and match BLAS/LAPACK libraries at the moment (also a challenge to support both LP64 and ILP64).

I have a WIP branch with a set of CI jobs where this is possible - at least different BLAS and LAPACK libraries (e.g., there's a job with `blas=blis lapack=openblas`): https://github.com/rgommers/scipy/tree/ci-blas. 

We had ILP64 support for OpenBLAS only in the `numpy.distutils` based build before. I was just in the process of resurrecting that support. The way it worked, unfortunately, is to link against both an LP64 and an ILP64 build of OpenBLAS. That is really ugly, and hard to even set up a test environment for. I'm not sure we should allow that again (it's unclear if anyone used that even). Both MKL and Accelerate provide a single library with two sets of symbols in them - that's much cleaner. OpenBLAS could be built like that as well in the future (reference BLAS/LAPACK too).

One thing I noticed in this PR is that the wrapper libraries are also built if no symbol prefix/suffix is needed. The way the old ILP64 support did that was a little different: it would write out a single header file (`blas64-prefix-defines.h`) and then rewrite all Fortran source files for build targets that depend on BLAS or LAPACK to include that header, if and only if a symbol suffix was needed. I was halfway through copying that approach into the Meson build. This approach seems to take about 20 seconds of extra build time on my macOS machine (2m 31s -> 2m 51s), so a ~10% increase. I suspect rewriting all the source files will be quite a bit faster, with fewer additional build targets.

I'd be curious about your opinions on the two topics in the two paragraphs above.

ilayn commented:
> In case we find ourselves in a situation like before again several years down the line, we can always drop support again. But right now there are only upsides to supporting Accelerate, and I don't see any reason not to.

True but keep in mind that we use LAPACK and NumPy does not and that's where the most of development happens. Reference BLAS has, for example, added many safety features as per Anderson's paper in 3.10 and I don't see any updates mentioned in any NumPy issues as far as I scrolled. So they have LAPACK 3.2.1 and 3.9.1 and who knows what now probably 3.20.1 next. 

Hence it is not that optimistic in my opinion and dropping and acquiring Accelerate support is really not very productive as they release a new version. But like I mentioned above, if this is not going to be a dependency I'm +1. If we are going to stop and wait until Accelerate catches up for years then -1.

thalassemia commented:
I took a look at the `numpy.distutils` approach. Making use of the preprocessor is a brilliant, virtually no-overhead way to handle the suffixes necessary for most ILP64 libraries. It's certainly far cleaner than this PR. 

For Accelerate specifically, I worry that Fortran's lack of case sensitivity will complicate things. The symbols provided by the new Accelerate are mixed case, so the linker would not be able to find them. I found a compiler flag for `ifort/ifx` (`-case as_is`) to force case sensitivity but not for modern `gfortran` (apparently the G77 version used to have `-fcase-preserve`). Adding this flag might break other things though.

Thanks for pointing out that the wrappers are being built whether they are needed or not. I'll change that so they are only built for Accelerate or, in the future, ILP64 builds.

Now that I've thought about it slightly more, I think my approach should support mixing BLAS/LAPACK as long as they share the same symbol names. The LP64 BLIS/OpenBLAS combo sounds fine in theory, but Accelerate cannot be matched with anything else because of its unique symbol names. I'm also worried that things might break when mixing ILP64 OpenBLAS and ILP64 MKL down the line given they use different compiler mangling conventions. 

I agree that it does not make much sense to support mixing LP64 and ILP64 libraries. In my testing, going all in on ILP64 by adding the Fortran compiler flags to make the default integer 64 bits only breaks `scipy.linalg._interpolative`, for which I needed to add a separate wrapper file. Surprisingly, the other Fortran modules just work with pretty minimal modifications (in fact, SuperLU, a C module, required the most tinkering to work).

It's great to hear PROPACK is getting converted to Python soon. Assuming the converted module accesses BLAS/LAPACK through `get_blas/lapack_funcs`, that should make my PR on that repo unnecessary and simplify ILP64 support in the future.

I'll mull this over some more. In the meantime, I'll break the other changes out into their own PRs. The removal of the `wrap_g77_abi_*` wrappers only works because they are replaced with C wrappers generated by my new scripts so I'll leave that here for now.

thalassemia commented:
I can't figure out what's causing the crash on the 32-bit build. After setting up the Docker container on my local machine, 0eac493 seemed to fix the segfault, but I also encountered test failures that did not occur here. It seems important not to mix Fortran and C wrappers when complex return values are involved. Also, when using C, the CBLAS ABI seems to be a safer bet than the Fortran one. It's confusing that issues only arise on a 32-bit OS.

rgommers commented:
> For Accelerate specifically, I worry that Fortran's lack of case sensitivity will complicate things

Hmm yes, that is a good point. The `blas64-prefix...` header generated all possible combinations of lower/upper-case symbols and mapped them to all-lowercase:
```C
#define ccopy ccopy_64
#define CCOPY ccopy_64
#define cCOPY ccopy_64
#define Ccopy ccopy_64
#define ccOPY ccopy_64
...
```
but that won't help with Accelerate if the name is `ccopy_$NEWLAPACK`.

> In my testing, going all in on ILP64 by adding the Fortran compiler flags to make the default integer 64 bits only breaks `scipy.linalg._interpolative`, for which I needed to add a separate wrapper file. Surprisingly, the other Fortran modules just work with pretty minimal modifications (in fact, SuperLU, a C module, required the most tinkering to work).

That's not too surprising, since they were the only two submodules that didn't support ILP64 before: https://github.com/scipy/scipy/pull/11193#issuecomment-568468741. We recently upgraded to SuperLU 6 though, and that's the first version that support ILP64 in the upstream build. I don't see any code changes for SuperLU in this PR though, other than adding the wrapper lib, so I'm curious - what was more difficult?

rgommers commented:
> It's great to hear PROPACK is getting converted to Python soon. Assuming the converted module accesses BLAS/LAPACK through `get_blas/lapack_funcs`, that should make my PR on that repo unnecessary and simplify ILP64 support in the future.

That may be months away, we should merge your PR and treat it as all the other Fortran code for now.

thalassemia commented:
I think I've finally figured out the x86 segfault. It seems that on 32-bit machines, `_Complex` and `npy_complex64` cannot be used interchangeably as a return type. To avoid segfaulting, the Fortran modules (ARPACK, PROPACK, etc) require the standard C99 `_Complex`, but the Cython modules (`_fblas`, `_flapack`, etc) require `npy_complex64`. `npy_complex64` is probably `struct {float real,imag;}` because Cython uses a struct for its complex type (`-DCYTHON_CCOMPLEX=0`) and converting between the two with union seems to work fine.

To address this, I brought back a separate wrapper file for the BLAS/LAPACK functions that return a complex value. By adding compile flags, I can change the complex type that is used for each build target.

I also added a flag to only generate wrappers for each BLAS/LAPACK function when `accelerate` is chosen, bringing the number of build targets and the build time back down for all other scenarios.

thalassemia commented:
> We recently upgraded to SuperLU 6 though, and that's the first version that support ILP64 in the upstream build. I don't see any code changes for SuperLU in this PR though, other than adding the wrapper lib, so I'm curious - what was more difficult?

Unfortunately, from what I can tell, the ILP64 support in SuperLU 6 is limited to their user-facing functions and internal array indexing. The underlying BLAS/LAPACK calls were not updated to support ILP64. To use an ILP64 BLAS/LAPACK, I wrote a Python script to replace all `int` and `unsigned int` across several files with `int64_t` and `uint64_t`, respectively. It worked on my machine, but this is almost certainly not a portable solution.

h-vetinari commented on scipy/_build_utils/src/complex_blas.c:
```diff
@@ -0,0 +1,81 @@
+
+#ifdef F_SRC
+void BLAS_FUNC(sladiv)(float *xr, float *xi, float *yr, float *yi, float *retr, float * reti);
+void F_FUNC(wcladiv, WCLADIV)(float_complex *ret, float_complex *x, float_complex *y){
+    BLAS_FUNC(sladiv)((float*)(x), (float*)(x)+1, \
+        (float*)(y), (float*)(y)+1, \
+        (float*)(ret), (float*)(ret)+1);
+}
```
Is the value of the imaginary components here irrelevant? I don't know `wcladiv`, but just based on the LAPACK [docs](https://netlib.org/lapack/explore-html/d5/db7/group__ladiv.html), just adding `+1` in the imaginary component is not equivalent.

thalassemia replied:
C99 `_Complex` types are apparently guaranteed to have the same representation and alignment requirements as an array type containing two elements of the corresponding real type, something I learned when working on this project. That means that this pointer arithmetic should work in theory. However, I just realized that the MSVC complex types do not conform to C99 and are instead structs just like in Cython and Numpy. This makes me wonder whether we would still see segfaults on 32-bit Windows. I don't have any way of testing that, but if it does, I could try declaring the arguments as real arrays of length 2.

h-vetinari commented on scipy/meson.build:
```diff
@@ -184,6 +205,10 @@ elif lapack_name == 'openblas'
 endif
+lapack_dep = declare_dependency(
+  dependencies: lapack,
+  compile_args: _args_blas
```
If we're going to use this for both BLAS & LAPACK, let's call it `_args_blas_lapack`.

h-vetinari commented on scipy/sparse/linalg/_propack/meson.build:
```diff
@@ -84,11 +84,6 @@ elements = [
 ]
 
 fargs_propack = ['-U_OPENMP']
-if use_g77_abi
-  # FIXME: define needs to be removed from PROPACK code, in favor of using
-  # `wcdotc` unconditionally, like is done in ARPACK.
-  fargs_propack += ['-DSCIPY_USE_G77_CDOTC_WRAP=1']
-endif
```
Can we split off all G77-related changes into a separate PR?

thalassemia replied:
Great idea! Now that the G77 wrappers are separate from `_generate_blas_wrappers.py`, this is finally possible #19855.

h-vetinari reviewed (commented):
Thanks for this PR! It's very cool to see this, and a really non-trivial thing to pull off. Well done!

That said, there are a few things I'd want to iterate on (resp. split off, like the G77 stuff). @rgommers probably has opinions™️too. ;-)

thalassemia commented:
Since this now depends on #19855, I have made it into a draft. When that PR is ready for merging, a large chunk of the diff here should disappear, and we can proceed with iterating.

thalassemia commented:
I noticed an issue with the `_fblas` and `_flapack` modules that are autogenerated by F2PY. The generated `PyInit` methods have code like the following for each `function` in the `*.pyf` signature file (example: `cdotu`):

```c
extern complex_float F_FUNC(cdotu,CDOTU)(void);
PyObject* o = PyDict_GetItemString(d,"cdotu");
tmp = F2PyCapsule_FromVoidPtr((void*)F_FUNC(cdotu,CDOTU),NULL);
PyObject_SetAttrString(o,"_cpointer", tmp);
Py_DECREF(tmp);
```

[Here is the relevant code](https://github.com/numpy/numpy/blob/7db1cc34e53e06ee2be0dab47d963789b075726f/numpy/f2py/rules.py#L458) in F2PY. [According to the docs](https://numpy.org/doc/stable/f2py/python-usage.html#fortran-type-objects), the `_cpointer` attribute is meant to be used as a callback argument pointing to the wrapped Fortran/C function.

The C code above is wrong because the wrapped function is not `F_FUNC(cdotu,CDOTU)` but `F_WRAPPEDFUNC(cdotu, CDOTU)`. This does not raise errors at build time because Accelerate defines `F_FUNC(cdotu, CDOTU)`, a macro that evaluates to `cdotu_` (old Accelerate uses symbols of the form `name_`, new Accelerate uses `name$NEWLAPACK`). However, compiling with Accelerate results in `_fblas.cdotu._cpointer` pointing to `F_FUNC(cdotu, CDOTU)` while `_fblas.cdotu` itself calls, in order:

1. `F_WRAPPEDFUNC(cdotu, CDOTU)`: This macro evaluates to `f2pywrapcdotu`, a Fortran wrapper function generated by F2PY ([`_fblas-f2pywrappers.f`](https://github.com/scipy/scipy/blob/51c5423f57d21ca193668b65287bcf00982a1aa7/scipy/linalg/meson.build#L61)).
2. `wcdotu_`: `f2pywrapcdotu` calls the external function `wcdotu`, which is mangled into `wcdotu_`. `wcdotu_` is defined in the [G77 ABI wrapper](https://github.com/scipy/scipy/blob/51c5423f57d21ca193668b65287bcf00982a1aa7/scipy/_build_utils/src/wrap_g77_abi.c#L50).
3. `cblas_cdotu_sub$NEWLAPACK`:  `wcdotu_` calls the C API for the new Accelerate. This is clearly different from `F_FUNC(cdotu,CDOTU)`, a G77 ABI function from the old Accelerate.


Making things more complicated, `F_WRAPPEDFUNC(cdotu, CDOTU)` is not a traditional function with a return value. Instead, it returns `void` and takes its "return" value as its first argument. To make this work, we would need to edit the C code above to something like:

```c
static complex_float _cpointer_cdotu(F_INT* n, complex_float* x, F_INT* incx, complex_float* y, F_INT* incy) {
    complex_float ret;
    F_WRAPPEDFUNC(cdotu, CDOTU)(ret, x, incx, y, incy);
    return ret;
}
PyObject* o = PyDict_GetItemString(d,"cdotu");
tmp = F2PyCapsule_FromVoidPtr((void*)_cpointer_cdotu,NULL);
PyObject_SetAttrString(o,"_cpointer", tmp);
Py_DECREF(tmp);
```

I doubt that anyone is even using the `_cpointer` attribute. Even so, this needs to be fixed or there will be build failures once we try to add support for BLAS libraries with symbol suffixes that do not also ship with suffix-free symbols like Accelerate.

All the comments by @h-vetinari still apply and are next on my to-do list. Thanks again for the detailed review here and on #19855!

thalassemia commented:
I found a way to fix the `_cpointer` issue by using `intent(c)` statements to prevent F2PY from generating Fortran wrappers and the erroneous boilerplate quoted above (7e791fe). This does not address the underlying F2PY issue ([discussion from 2014](https://github.com/scipy/scipy/pull/4021#issuecomment-56930397)).

Removing the Fortran wrappers reintroduces the x86 segfaults caused by incompatibility between struct (e.g. `{float real, imag;}`) and C99 complex return types. Cython uses struct complex types due to `-DCYTHON_CCOMPLEX=0`, and F2PY uses struct complex types by default. However, our G77 wrappers and pure Fortran modules (e.g. PROPACK) use C99 complex types. To fix this, I wrote `wrap_g77_fortran_intermediate.f` to provide Fortran subroutine wrappers for the affected BLAS/LAPACK functions. This file is bundled into a `*.a` with the G77 ABI wrappers (or dummy) and provides symbols of the form `{name}wrp_` for Cython/F2PY modules (6404580).

ilayn commented:
I think that is an acceptable workaround. Frankly, I can't wait to convert PROPACK and be done with this saga.

thalassemia commented:
I've addressed @h-vetinari's comments by adding more inline documentation and refactoring both `_generate_blas_wrapper.py` and `_generate_pyx.py`. 

Previously, to fix the x86 segfaults,  I let `_generate_pyx.py` and F2PY create Fortran subroutine wrappers around functions with complex return values. Unfortunately, I had to forgo the F2PY-generated wrappers to fix the `_cpointer` issue described in my last comment. In place of these wrappers, I've appended equivalent C wrappers to `wrap_g77_abi.c` and `wrap_dummy_g77_abi.c`. This let me remove all the Fortran wrapper generation code in `_generate_pyx.py` and link both Cython and F2PY modules against the G77 (or dummy) wrapper library.

h-vetinari reviewed (commented):
Thank you for the updates @thalassemia!

I see now that many of the structural things I had criticised were actually taken over directly from `scipy/linalg/_generate_pyx.py` - which was a very reasonable thing to do, so I apologise for that. Thanks a lot for making the effort to also clean up that structure in the other file!

That said, I'm now a big concerned about the duplication between the two files `scipy/_build_utils/_generate_blas_wrapper.py` and `scipy/linalg/_generate_pyx.py`. For example, `read_signatures` is fully duplicated, as are several type mappings, contextual things (header guards, special cases) etc.

I would like to break out common infrastructure/definitions into (for example) `scipy/_build_utils/_wrappers_common.py` or something like that -- at least things like `read_signatures`, `c_types`, `fortran_types`,  `npy_types`, `wrapped_funcs` (perhaps renamed to `complex_return_funcs`), `cpp_guard_{begin,end}`, and import those functions/quantities from the other two files.

There's more unification opportunities as well between the two files. For example, the `generate_*` functions in `scipy/_build_utils/_generate_blas_wrapper.py` could also defer the writing of the files to `make_all`, after similarly doing an `all_newer` check.

On a higher level, I'm wondering if we can distinguish the different generation targets (pxy, pxd, c header, wrapper) could share a generation function, with the only difference being the template (or f-string) being filled? Probably one function is too optimistic but I feel we do a bunch of similar processing.

In general, for these generation functions, I'd like to use similar names (e.g. `generate_*` or `create_*`), and use them in recognizable patterns. E.g.
```python
COMMENT_TEXT = [f"This file was generated by {__FILE__}.\n",
                "Do not edit this file directly.\n"]
def generate_decl_c(name, return_type, argnames, argtypes, accelerate):
    """Create C header declarations for Cython to import."""
    # ...
def generate_decl_pyx(name, return_type, argnames, argtypes, accelerate, header_name):
    """Create Cython declaration for BLAS/LAPACK function."""
    # ...
def generate_decl_pxd(name, return_type, argnames, argtypes, accelerate):
    # even if it's exceedingly trivial
def generate_decl_wrapper(name, return_type, argnames, argtypes, accelerate):
    # renamed from create_wrapper

def generate_file_content_pyx(sigs, lib_name, accelerate):
    """Generate pyx file with BLAS/LAPACK func declarations and tests."""
    # add appropriately escaped COMMENT_TEXT before premable too (same below)
    # ... [generate_decl_pxy(**sig) for sig in sigs)]
    return "\n".join(content)
def generate_file_content_pxd(sigs, lib_name, accelerate):
    """Create pxd header file for generated pyx."""
    # ... [generate_decl_pxd(**sig) for sig in sigs)]
    return "\n".join(content)
def generate_file_content_header(sigs, lib_name, accelerate):
    """Generate complete C header file with declarations and includes."""
    # ... [generate_decl_c(**sig) for sig in sigs)]
    return "\n".join(content)
def generate_file_content_wrappers(sigs, lib_name, accelerate):
    """Generate file content for wrappers"""
    # ... [generate_decl_wrapper(**sig) for sig in sigs)]
    # here we probably want to return a dict {filename: content}

def make_all(...):  # for linalg case
    src_files = ...
    dst_files = ...
    if all_newer(src_files, dst_files): return

    blas_sigs = read_signatures(blas_sigs)
    blas_pyx = generate_file_content_pyx(blas_sigs, "BLAS", accelerate)
    blas_pxd = generate_file_content_pxd(blas_sigs, "BLAS", accelerate)
    blas_header = generate_file_content_header(blas_sigs, "BLAS", accelerate)
    lapack_sigs = read_signatures(lapack_sigs)
    lapack_pyx = generate_file_content_pyx(blas_sigs, "LAPACK", accelerate)
    lapack_pxd = generate_file_content_pxd(blas_sigs, "LAPACK", accelerate)
    lapack_header = generate_file_content_header(blas_sigs, "LAPACK", accelerate)
    to_write = {"cython_blas.pyx": blas_pyx, "cython_blas.pxd": blas_pxd, ...}
    for file_name, content in to_write.items():
        with open(os.path.join(outdir, file_name), "w") as f:
            f.write(content)

def make_all(...):  # for wrapper case
    src_files = ...
    dst_files = ...
    # may consider moving this skip below the content generation, since that gets us filenames dynamically
    if all_newer(src_files, dst_files):
        return

    blas_sigs = read_signatures(blas_sigs)
    lapack_sigs = read_signatures(lapack_sigs)
    # *_wrappers are already a dict {file_name: content}
    blas_wrappers = generate_file_content_wrappers(blas_sigs, 'BLAS', accelerate)  
    lapack_wrappers = generate_file_content_wrappers(lapack_sigs, 'BLAS', accelerate)
    to_write = dict(**blas_wrappers, **lapack_wrappers)
    for file_name, content in to_write.items():
        with open(os.path.join(outdir, file_name), "w") as f:
            f.write(content)
```
Note that this is a pretty quick draft from looking at this PR for a bit. It's possible we may still want to do this differently, but at least it should make clear that I think we should try to use common patterns in all those cases.

I'd suggest we iterate a bit in this PR (or another) how this should look like. Then I'd like to split out refactoring the existing `scipy/linalg/_generate_pyx.py` into a separate PR, and finally, introduce this on top. More or less independently, I think we can split off the `_cpointer` issue into a separate PR as well.

Sorry for this long (and demanding) review, and thanks a lot for your work on this so far! If you want I can also participate more directly in the iterations towards a cleaner generation infrastructure.

thalassemia commented:
Thanks again for the feedback @h-vetinari! I refactored a bit more and unified as much as I could in `_build_utils/_wrappers_common.py`. To import the unified code module in `linalg/_generate_pyx.py`, I had to use an ugly `sys.path` hack that the linter is unhappy with.

thalassemia committed:
TST: Add macOS Accelerate CI test

[skip circle]

andyfaff commented:
I can help with adding the CI entry. Does this PR successfully test locally?

thalassemia commented:
Thanks @andyfaff. I think I have the CI entry set up the way I want now (did not realize that `macos-latest` was `macos-12` even though `macos-13` and `macos-14` are available). I have never run the full test suite locally before. Seems that using Accelerate results in 2 failures:
- `TestInterpN.test_matrix_input[quintic]`: relative error 5.4e-7 > relative tolerance 1e-7
- `TestInterpN.test_matrix_input[quintic]`: numerical issues that I am not qualified to address

andyfaff commented:
I think it's probably better to test on `macos-14`, as eventually wheels will be deployed on macos>=14.

lucascolley commented:
btw @thalassemia , feel free to save some CI resources by adding `[skip circle]` to skip the docs workflows as outlined in http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping whilst iterating here - not a requirement at all, but we appreciate it!

thalassemia commented:
I can reproduce the one Accelerate test failure (`TestSqrtM.test_gh17918`) locally. The relevant code was recently modified in #20212. Reverting that commit fixes the issue.

lucascolley commented:
> I can reproduce the one macOS 14 M1 Accelerate test failure (TestSqrtM.test_gh17918) locally

cc @ev-br, maybe a slight bump needed to accommodate for Accelerate?

ev-br commented:
Thanks for the ping Lucas.

If somebody has the right hardware + cycles to chase it down: 

```
$ git diff
diff --git a/scipy/linalg/_matfuncs_sqrtm.py b/scipy/linalg/_matfuncs_sqrtm.py
index 9c84ca573b..5655726d07 100644
--- a/scipy/linalg/_matfuncs_sqrtm.py
+++ b/scipy/linalg/_matfuncs_sqrtm.py
@@ -170,6 +170,7 @@ def sqrtm(A, disp=True, blocksize=64):
     if blocksize < 1:
         raise ValueError("The blocksize should be at least 1.")
     keep_it_real = np.isrealobj(A)
+    breakpoint()
     if keep_it_real:
         T, Z = schur(A)
         d0 = np.diagonal(T)

```

and then locally I see

```
173  	    breakpoint()
174  	    if keep_it_real:
175  	        T, Z = schur(A)
176  ->	        d0 = np.diagonal(T)
177  	        d1 = np.diagonal(T, -1)
178  	        eps = np.finfo(T.dtype).eps
179  	        needs_conversion = abs(d1) > eps * (abs(d0[1:]) + abs(d0[:-1]))
180  	        if needs_conversion.any():
181  	            T, Z = rsf2csf(T, Z)
(Pdb) n
> /home/br/repos/scipy/scipy/build-install/lib/python3.10/site-packages/scipy/linalg/_matfuncs_sqrtm.py(177)sqrtm()
-> d1 = np.diagonal(T, -1)
(Pdb) n
> /home/br/repos/scipy/scipy/build-install/lib/python3.10/site-packages/scipy/linalg/_matfuncs_sqrtm.py(178)sqrtm()
-> eps = np.finfo(T.dtype).eps
(Pdb) p d0
array([ 0.06, 17.92,  0.06,  0.06,  0.06,  0.06,  0.06,  0.06,  0.06,
        0.06,  0.06,  0.06,  0.06,  0.06,  0.06,  0.06,  0.06,  0.06,
        0.06])
(Pdb) p d1
array([ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00, -4.77048956e-18,
        0.00000000e+00,  0.00000000e+00,  0.00000000e+00,  0.00000000e+00,
        0.00000000e+00,  0.00000000e+00,  0.00000000e+00,  0.00000000e+00,
        0.00000000e+00,  0.00000000e+00,  0.00000000e+00,  0.00000000e+00,
        0.00000000e+00,  0.00000000e+00])
```

That said, I guess a conditional skip is fine if this otherwise distracts from the main push of this PR.

rgommers commented:
The `sqrtm` failure is not Accelerate-specific. Here it is failing in the wheel build jobs for macOS x86-64 with OpenBLAS 0.3.21: https://github.com/scipy/scipy/actions/runs/8201161926/job/22429344365.

It's blocking rebuilding the nightlies, which we need. So I'll mark it as xfail in `main` now and will open a separate issue to fix it up.

thalassemia commented:
@ev-br Here is the requested output from my M2 macOS 14 machine with the latest build of this branch:

```sh

```

h-vetinari commented:
Thanks a lot for the updates @thalassemia! I didn't go over it with a fine-toothed comb, and I doubt doubt that there are details that could be polished if people feel find something they feel strongly about, but from my POV, this is now a very nice improvement to the generation infrastructure, completely aside from adding the capability to handle macOS 13.3+. 🙏

h-vetinari commented:
In an ideal world, we'd split off the changes to the existing infrastructure into a separate PR and then add support for 13.3+ on top, but it's not a requirement from my side. It just makes history digging / bisecting etc. easier if we can split off refactoring from new feature additions. 🙃

tylerjereddy commented:
I'd like to branch for SciPy `1.13.0` around Sunday (17th). Let me know where this stands/should stand with respect to merging by then vs. delaying to merge after branching.

thalassemia commented:
@tylerjereddy This depends on gh-20232 but will have a much smaller diff once that is merged. I think it would be nice if @rgommers could review this in its final form before merging. To give more time for that, it would probably be safer to hold off on merging until after the branch.

h-vetinari commented:
I definitely want to get this into 1.13 (pending at least some review by @rgommers).

rgommers commented:
It probably doesn't matter too much whether this gets into 1.13.0 or not, since we won't be able to ship Accelerate-based wheels in time, and conda-forge also doesn't have support yet for new Accelerate. But I'll try to re-review asap.

thalassemia commented:
I rebased this onto gh-20232 and squashed everything down to two commits for clarity. The other commits were either iterative or split into separate PRs (all linked in the updated description).

rgommers commented on scipy/meson.build:
```diff
@@ -148,6 +148,23 @@ fortranobject_dep = declare_dependency(
+_args_blas_lapack = []
+accelerate_flag = []
+generate_blas_wrappers = false
+if blas_name == 'accelerate' and macOS13_3_or_later
```
Old Accelerate is not supported, so I suggest changing this to:
```meson
if blas_name == 'accelerate'
  if not macOS13_3_or_later
    error('macOS Accelerate is only supported on macOS >=13.3')
```

rgommers commented on scipy/_build_utils/src/npy_cblas.h:
```diff
@@ -26,6 +26,17 @@ enum CBLAS_SIDE {CblasLeft=141, CblasRight=142};
 #define CBLAS_INDEX size_t  /* this may vary between platforms */
 
+#ifdef ACCELERATE_NEW_LAPACK
+    #if __MAC_OS_X_VERSION_MAX_ALLOWED < 130300
```
This branch should never be reachable, right, since `ACCELERATE_NEW_LAPACK` should only be defined on macOS >=13.3?

rgommers reviewed (commented):
Thanks! Much easier to review now. I haven't played with this version yet, but a first few comments about Accelerate: we'd like to add support for new Accelerate, but (unlike NumPy) do not support the old Accelerate (see https://github.com/scipy/archive/blob/main/wiki/dropping-accelerate-support.md). So we should not try here, but rather error out on macOS <13.3

h-vetinari commented on scipy/_build_utils/_generate_blas_wrapper.py:
```diff
@@ -0,0 +1,126 @@
+
+CURR_DIR = os.path.dirname(os.path.abspath(__file__))
+LINALG_DIR = os.path.abspath(os.path.join(CURR_DIR, "..", "linalg"))
+comments = f"""/*
+This file was generated by {os.path.basename(__file__)}.
+Do not edit this file directly.
+*/"""
```
I can't see this being inserted anywhere. I'd also like to call it `COMMENT_TEXT` (or `C_COMMENT` or whatever), like in `_generate_pyx.py`.

h-vetinari commented on scipy/_build_utils/_generate_blas_wrapper.py:
```diff
@@ -0,0 +1,126 @@
+"""
+
+
+def generate_file_wrapper(sigs, lib_name, accelerate, outdir):
```
Let's add a short docstring, and that this returns a dictionary of `{file_path: file_content}`.

h-vetinari reviewed (commented):
Just some remaining nits...

h-vetinari commented on scipy/_build_utils/_generate_blas_wrapper.py:
```diff
@@ -0,0 +1,126 @@
+    # Complex-valued functions have hard-coded wrappers in G77 ABI wrappers
+    if name in WRAPPED_FUNCS:
+        return ""
+    # lsame and dcabs1 are missing from new Accelerate
+    # Use old symbols that do not require a wrapper
+    if accelerate and name in ['lsame', 'dcabs1']:
+        return ""
```
This is coupled with the updated handling in `get_blas_macro_and_name`. Perhaps we want to introduce `SYMBOLS_MISSING_IN_NEWLAPACK` to make this link more self-evident?

tylerjereddy commented:
Branching is imminent, so I've bumped the milestone. If there's a compelling benefit to having this in `1.13.0` we could discuss a backport, but it doesn't sound like there is.

h-vetinari commented on scipy/_build_utils/_generate_blas_wrapper.py:
```diff
@@ -0,0 +1,132 @@
+
+CURR_DIR = os.path.dirname(os.path.abspath(__file__))
+LINALG_DIR = os.path.abspath(os.path.join(CURR_DIR, "..", "linalg"))
+C_COMMENT = f"""/*
+This file was generated by {os.path.basename(__file__)}.
+Do not edit this file directly.
+*/"""
```
Since the final pieces are joined together using `"".join()`, we should end this with a newline, or even two to be clear.
```suggestion
C_COMMENT = f"""/*
This file was generated by {os.path.basename(__file__)}.
Do not edit this file directly.
*/\n\n"""
```

h-vetinari reviewed (approved):
Just a formatting nit, otherwise LGTM! Very nice!

thalassemia committed:
CI: pip instead of Conda for accelerate test

Avoids bug in Clang 16 (gh-20208)
[skip circle]

thalassemia commented:
After sitting on this for a while, I feel ready to merge. Although it pains me to dramatically increase the number of build targets just to support Accelerate (~1400 -> ~3100), the build times seem to be under control as evidenced by my M1 Accelerate GHA taking in 10 minutes while the M1 OpenBLAS test took 11 minutes. I could not figure out how to get the macOS linker to exclude unused symbols when consolidating all the wrappers into a single compilation target. I wonder if this might be possible on other platforms with other linkers (will explore in a future ILP64 PR).

CC @rgommers for any final comments before merging. Thanks again for all the reviews these past few months @h-vetinari!

rgommers commented:
Thanks @thalassemia. Working on going through this once more and exercising it a bit.

> the build times seem to be under control as evidenced by my M1 Accelerate GHA taking in 10 minutes while the M1 OpenBLAS test took 11 minutes.

Agreed. I see a 10%-15% increase in build time on my M1 Macbook Pro. Which I agree is acceptable. Tracing shows that the added targets are adding a ~20 sec load of small file compiles at the start of the build, which makes sense:

<img width="1208" alt="image" src="https://github.com/scipy/scipy/assets/98330/63428ef0-182a-4c91-aeb9-648bdb2b34c9">

The build time increase will only occur with Accelerate, so that's all good.

rgommers commented:
I checked the binary size change: this approach is very lean, 54 kb extra due to the extra static libraries. Size grows from 21.78 MB to 21.84 MB for a local build. The `macosx_arm64` wheels on PyPI are 30.3 MB, so this is (as expected) a major win - we can start shipping `macosx_14_0_arm64` wheels that are almost a third smaller than the `macosx_12_0_arm64` ones.

> I could not figure out how to get the macOS linker to exclude unused symbols when consolidating all the wrappers into a single compilation target. I wonder if this might be possible on other platforms with other linkers (will explore in a future ILP64 PR).

I think it's possible with linker flags like `-Wl,--gc-sections` perhaps, but not in a portable way. Enabling LTO may help - see https://github.com/scipy/scipy/issues/16098#issuecomment-1887155182.

rgommers commented on scipy/spatial/meson.build:
```diff
@@ -40,7 +40,7 @@ py3.extension_module('_qhull',
   ],
   link_args: version_link_args,
-  dependencies: [np_dep],
+  dependencies: [lapack_dep, np_dep],
```
This shouldn't be necessary, and removing it works for me locally. If it is needed for something, that'd be a potential problem with `cython_lapack` usage. Is this a left-over, or do you still see a need for this?

thalassemia replied:
Good catch! This was a holdover from my ILP64 experiments. `qhull_misc.h` includes `npy_cblas.h` for `CBLAS_INT`, which is a 64-bit integer if `HAVE_BLAS_ILP64` is defined. I plan to add this as an argument to `lapack_dep` and will reintroduce this change in a separate PR.

rgommers commented on scipy/meson.build:
```diff
@@ -184,6 +208,10 @@ elif lapack_name == 'openblas'
 endif
+lapack_dep = declare_dependency(
+  dependencies: lapack,
+  compile_args: _args_blas_lapack
```
It's a little concerning that some targets need `link_with: blas_lapack_wrapper_lib` and some don't. It would probably be nicer to include that `link_with` here, then using `lapack_dep`/`blas_dep` anywhere else will be robust.

thalassemia replied:
Another really good catch! I checked `scipy.linalg._interpolative`, one of the targets where this was missing, and realized that the only reason nothing broke was because Accelerate provides the updated `$NEWLAPACK`-suffixed symbols alongside the suffix-less old symbols. To prevent any wrapper omissions going forward, I added `link_with: [g77_abi_wrappers, blas_lapack_wrapper_lib]` to `lapack_dep`/`blas_dep`.

thalassemia committed:
BLD: Tweak symbol visibility so macOS linker can strip unused symbols

thalassemia committed:
BLD: Include BLAS/LAPACK link libraries in Meson dependency

[skip circle]
[skip cirrus]

thalassemia commented:
Good news! I think I found a way to put every wrapper in a single file and let the linker exclude unused symbols. Previously, the linker was forced to include otherwise unused symbols simply because all exported symbols are global by default. Now, by explicitly marking my wrapper symbols as hidden, the linker can remove any symbols that are not directly used within each individual module.

Importantly, this PR only adds the flags needed to make this work in macOS, since that is the only platform with Accelerate. In a future ILP64 PR using this framework, I'll need to add additional compiler-specific Meson flags to achieve the same result (`-ffunction-sections -fdata-sections -Wl,--gc-sections` for GCC and `/Gy` for MSVC)

h-vetinari commented on scipy/_build_utils/src/wrap_dummy_g77_abi.c:
```diff
@@ -39,69 +39,92 @@ typedef double _Complex double_complex;
-float_complex F_FUNC(wcdotc,WCDOTC)(CBLAS_INT *n, float_complex *cx, \
-        CBLAS_INT *incx, float_complex *cy, CBLAS_INT *incy){
+/* Let linker strip unused symbols from final binaries */
+#if __GNUC__ >= 4
+#define HIDDEN_SYMBOL  __attribute__ ((visibility ("hidden")))
+#else
+#define HIDDEN_SYMBOL
+#endif
```
Clang understands `__attribute__ ((visibility ("hidden")))`, though just from the condition, it's not obvious that this branch will also be taken there.

As it happens, clang will also [populate](https://clang.llvm.org/docs/UsersManual.html#cmdoption-fgnuc-version) `__GNUC__`, and while the default value is not documented (from what I saw), it just so happens to be `4`, see:
* https://github.com/llvm/llvm-project/blob/llvmorg-18.1.3/clang/lib/Driver/ToolChains/Clang.cpp#L6892
* https://github.com/llvm/llvm-project/blob/llvmorg-18.1.3/clang/lib/Frontend/CompilerInvocation.cpp#L3838
* https://github.com/llvm/llvm-project/blob/llvmorg-18.1.3/clang/lib/Frontend/InitPreprocessor.cpp#L794

In other words, the `4` here is load-bearing for clang support, and we cannot increment it to our lowest supported GCC version, for example. We definitely need to document this.

thalassemia replied:
Oh wow, I had no idea Clang defined such a low value for `__GNUC__`. I copied these defines from [the GCC wiki](https://gcc.gnu.org/wiki/Visibility) and just assumed things would be fine since they worked on my local machine. I moved this into `npy_cblas.h` as you suggested and added a comment about which compilers this targets and why it exists.

h-vetinari commented on scipy/_build_utils/src/wrap_g77_abi.c:
```diff
@@ -39,41 +39,54 @@ typedef double _Complex double_complex;
 
-float_complex F_FUNC(wcdotc,WCDOTC)(CBLAS_INT *n, float_complex *cx, \
-        CBLAS_INT *incx, float_complex *cy, CBLAS_INT *incy){
+/* Let linker strip unused symbols from final binaries */
+#if __GNUC__ >= 4
+#define HIDDEN_SYMBOL  __attribute__ ((visibility ("hidden")))
+#else
+#define HIDDEN_SYMBOL
+#endif
```
I also don't like defining this in three places.

h-vetinari commented on scipy/_build_utils/_generate_blas_wrapper.py:
```diff
@@ -61,31 +61,29 @@ def generate_decl_wrapper(name, return_type, argnames, argtypes, accelerate):
+    # Let linker strip symbols from final binaries
+    file_text.append("""
+#if __GNUC__ >= 4
+#define HIDDEN_SYMBOL  __attribute__ ((visibility ("hidden")))
```
From what I can tell, we could put this in `npy_cblas.h`, which is included directly in the other two files, and gets picked up here through:
https://github.com/scipy/scipy/blob/438e88f5e1752f49b5c760e15c94c48f26d83e4d/scipy/_build_utils/_wrappers_common.py#L37-L40

h-vetinari commented on scipy/_build_utils/src/npy_cblas.h:
```diff
@@ -15,6 +15,18 @@ extern "C"
+/*
+ * GCC and Clang require functions from static libraries to have this attribute
+ * in order for the linker to strip them if they are unused during linkage of
+ * dynamic libraries. We require GCC > 4 and Clang defines __GNUC__ as 4. MSVC
+ * does not require this macro because it makes all symbols hidden by default.
```
A bit more details here, so that future readers don't have to go digging into the PR comments.
```suggestion
 * dynamic libraries, see https://gcc.gnu.org/wiki/Visibility.
 * This is supported since GCC 4.1 (though our lowest supported version is much
 * higher already), which we keep using here since Clang defines __GNUC__ as 4
 * (since at least Clang 2.7 and up to the most recent Clang 18), see
 * https://clang.llvm.org/docs/UsersManual.html#cmdoption-fgnuc-version
 * MSVC does not require this macro because it hides all symbols by default.
```

thalassemia replied:
Did away with this complexity by using the `gnu_symbol_visibility` Meson option.

h-vetinari commented on meson.build:
```diff
@@ -84,6 +84,8 @@ endif
 if host_machine.system() == 'darwin' and cc.has_link_argument('-Wl,-ld_classic')
   # New linker introduced in macOS 14 not working yet, see gh-19357 and gh-19387
   add_project_link_arguments('-Wl,-ld_classic', language : ['c', 'cpp', 'fortran'])
+  # Allow linker to strip unused symbols
+  add_project_link_arguments('-Wl,-dead_strip', language : ['c', 'cpp', 'fortran'])
 endif
```
Not sure this is necessary, but I note that the guard check here explicitly tests for availability of the existing linker argument before adding it. Perhaps we should do the same for `-Wl,-dead_strip`?
```suggestion
if host_machine.system() == 'darwin'
  if cc.has_link_argument('-Wl,-ld_classic')
    # New linker introduced in macOS 14 not working yet, see gh-19357 and gh-19387
    add_project_link_arguments('-Wl,-ld_classic', language : ['c', 'cpp', 'fortran'])
  endif
  if cc.has_link_argument('-Wl,-dead_strip')
    # Allow linker to strip unused symbols
    add_project_link_arguments('-Wl,-dead_strip', language : ['c', 'cpp', 'fortran'])
  endif
endif
```

rgommers reviewed (commented):
> Previously, the linker was forced to include otherwise unused symbols simply because all exported symbols are global by default. 

Regarding hidden symbols, it shouldn't be needed to define those manually or use `__attribute__('hidden')` directly I suspect. Instead, the `gnu_symbol_visibility` keyword to [static_library](https://mesonbuild.com/Reference-manual_functions.html#static_library) can be used to do this automatically - and we wouldn't have to worry about things like Clang defining `__GNU__`. Did you consider this?

rgommers commented:
Re `gnu_symbol_visibility`, I just realized that we should be using that anyway on all static libraries - see gh-20477.

thalassemia committed:
BLD: Set Meson gnu_symbol_visibility to hidden for BLAS/LAPACK wrappers

[skip circle]
[skip cirrus]

thalassemia commented:
Did not know about the `gnu_symbol_visibility` option. That is a lot cleaner and works for me locally using either OpenBLAS or Accelerate. I updated my PR to use that.

thalassemia commented on scipy/sparse/linalg/_dsolve/meson.build:
```diff
@@ -196,6 +196,7 @@ superlu_lib = static_library('superlu_lib',
     'SuperLU/SRC/zutil.c'
   ],
   c_args: _superlu_lib_c_args,
+  dependencies: blas_dep,
```
Looking at this more closely, I think this line is unnecessary. Even if the BLAS dependency is missing when compiling the static library, it is included when compiling the extension module that links with this library. The same applies to the BLAS/LAPACK dependencies I added to the ARPACK static library. I'll clean those up for clarity.

thalassemia committed:
MAINT: Remove unnecessary BLAS/LAPACK deps

rgommers reviewed (approved):
LGTM now. Time to give this a go. Thanks a lot @thalassemia! And thanks @h-vetinari for all the review effort!

h-vetinari commented:
Congratulations @thalassemia and thanks for splitting various pieces off from this big effort! This definitely deserves a release note - if you like, please edit one into [the current draft](https://github.com/scipy/scipy/wiki/*Release-Note-Entries-for-SciPy-1.14.0-(asterisk-makes-this-appear-at-the-top)).

andyfaff commented:
Yes, congratulations @thalassemia, a big effort! Do we know how much the wheel size will reduce with not having to bundle openblas?

andyfaff commented:
> Do you think it makes sense to have our macOS wheels use Accelerate by default?

It's only useful for a certain mac version and above. Below a certain version I think there are issues with the library. I think the version is 14, but I can't remember off the top of my head.

This means two sets of wheels have to be made. One set for version > 10.X that bundles openblas and one set for version >14.x that doesn't. numpy is already doing this, but start from a lower macOS version.

rgommers commented:
Yes indeed, we should start shipping extra wheels built against Accelerate for macOS >=14.0 for the next release, just like is done for numpy 2.0.0. They will be smaller and faster.

rgommers commented:
This may be a good place to record that macOS 15 will include another update to Accelerate - it upgrade to LAPACK 3.11.0 with no change in symbol names. The extra LAPACK APIs will be available if the macOS deployment target is >= 15.0.
